pkg/segment/query/processor/statscommand.go

```diff
@@ -18,22 +18,151 @@
+	if inputIQR == nil {
+		if p.searchResults != nil {
+			inputIQR = iqr.NewIQR(p.qid)
+			return p.extractGroupByResults(inputIQR)
+		}
+		return nil, io.EOF
+	}
+
+	numOfRecords := inputIQR.NumberOfRecords()
+
+	if p.bucketKeyWorkingBuf == nil {
+		p.bucketKeyWorkingBuf = make([]byte, len(p.options.GroupByRequest.GroupByColumns)*utils.MAX_RECORD_SIZE)
+	}
+
+	if p.searchResults == nil {
+		p.options.GroupByRequest.BucketCount = int(utils.QUERY_MAX_BUCKETS)
+		p.options.GroupByRequest.IsBucketKeySeparatedByDelim = true
+		searchResults, err := segresults.InitSearchResults(uint64(numOfRecords), &structs.QueryAggregators{GroupByRequest: p.options.GroupByRequest}, structs.GroupByCmd, inputIQR.GetQID())
+		if err != nil {
+			return nil, toputils.TeeErrorf("stats.Process.processGroupByRequest: cannot initialize search results; err=%v", err)
+		}
+		p.searchResults = searchResults
+	}
+
+	blkResults := p.searchResults.BlockResults
+
+	measureInfo, internalMops := blkResults.GetConvertedMeasureInfo()
+	measureResults := make([]utils.CValueEnclosure, len(internalMops))
+
+	for i := 0; i < numOfRecords; i++ {
+		record := inputIQR.GetRecord(i)
+
+		// Bucket Key
+		bucketKeyBufIdx := 0
+
+		for idx, cname := range p.options.GroupByRequest.GroupByColumns {
+			if idx > 0 {
+				copy(p.bucketKeyWorkingBuf[bucketKeyBufIdx:], utils.BYTE_TILDE)
+				bucketKeyBufIdx += utils.BYTE_TILDE_LEN
+			}
+
+			cValue, err := record.ReadColumn(cname)
+			if err != nil {
+				p.errorData.readColumns[cname] = struct{}{}
+				copy(p.bucketKeyWorkingBuf[bucketKeyBufIdx:], utils.VALTYPE_ENC_BACKFILL)
+				bucketKeyBufIdx += 1
+			} else {
+				bytesVal := cValue.ConvertToBytesValue()
+				copy(p.bucketKeyWorkingBuf[bucketKeyBufIdx:], bytesVal)
+				bucketKeyBufIdx += len(bytesVal)
+			}
+		}
+
+		for cname, indices := range measureInfo {
+			cValue, err := record.ReadColumn(cname)
+			if err != nil {
+				p.errorData.readColumns[cname] = struct{}{}
+				cValue = &utils.CValueEnclosure{CVal: utils.VALTYPE_ENC_BACKFILL, Dtype: utils.SS_DT_BACKFILL}
+			}
+
+			for _, idx := range indices {
+				measureResults[idx] = *cValue
+			}
+		}
+		blkResults.AddMeasureResultsToKey(p.bucketKeyWorkingBuf[:bucketKeyBufIdx], measureResults, "", false, inputIQR.GetQID())
+	}
+
+	return nil, nil
+}
+
+func (p *statsProcessor) extractGroupByResults(iqr *iqr.IQR) (*iqr.IQR, error) {
+	if p.searchResults == nil {
+		return iqr, nil
+	}
+
+	// load and convert the bucket results
+	_ = p.searchResults.GetBucketResults()
+
+	bucketHolderArr, measureFuncs, aggGroupByCols, _, bucketCount := p.searchResults.GetGroupyByBuckets(int(utils.QUERY_MAX_BUCKETS))
+
+	err := iqr.AppendStatsResults(bucketHolderArr, measureFuncs, aggGroupByCols, bucketCount)
+	if err != nil {
+		return nil, toputils.TeeErrorf("stats.Process.extractGroupByResults: cannot append stats results; err=%v", err)
+	}
+
+	p.logErrors()
+
+	return iqr, io.EOF
```

Comment:
Should you return an actual error instead of `EOF` if you got an error?

Reply:
In that case, I should return as soon as I found an error. Instead of tracking them. 

But for Group by for example `col1`, `col2`. Col2 may not exist in the first segment or batch of results but it might exist in the next batch.

So for that reason, I am just tracking the errors and logging them. And backfilling values for the column that throws an error.

Reply:
Ok, then should they be logged as Warn/Info? I think currently you're logging them as errors